storage/replica_command.go

```diff
@@ -1407,11 +1407,17 @@ func (r *Replica) TruncateLog(
 		}); err != nil {
 		return reply, err
 	}
-	tState := roachpb.RaftTruncatedState{
+	tState := &roachpb.RaftTruncatedState{
 		Index: args.Index - 1,
 		Term:  term,
 	}
-	return reply, engine.MVCCPutProto(ctx, batch, ms, keys.RaftTruncatedStateKey(r.RangeID), roachpb.ZeroTimestamp, nil, &tState)
+
+	batch.(engine.Batch).Defer(func() {
```

Comment:
`Replica.executeCmd` takes an `engine.ReadWriter` because `Replica.executeBatch` does because sometimes that method is not called with a batch. But, we always pass an `engine.Batch` to `Replica.executeBatch` when executing a batch containing write methods. So the short answer is that it isn't trivial to plumb the type information. The longer answer is that we should always be passing an `engine.Batch` to `Replica.executeBatch`, we just need an `engine.ReadOnlyBatch` to make this possible. See #6803.